c/rust/src/errors/error.rs

```diff
@@ -30,22 +33,31 @@ impl From<playground::ToyError> for Error {
     }
 }
 
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Self {
+        Error::IOError(err)
+    }
+}
+
 impl Error {
-    fn domain(self) -> ErrorDomain {
+    fn domain(&self) -> ErrorDomain {
         match self {
             Error::PlaygroundToyError(_) => ErrorDomain::PlaygroundToyError,
+            Error::IOError(_) => ErrorDomain::IOError,
         }
     }
 
-    fn code(self) -> u32 {
+    fn code(&self) -> i32 {
         match self {
-            Error::PlaygroundToyError(e) => e as u32,
+            Error::PlaygroundToyError(e) => *e as i32,
+            Error::IOError(e) => e.raw_os_error().unwrap_or(-1),
```

Comment:
Not a big fan of making error codes signed. As far as I can tell OS errors are always >=1, so can we use 0 instead of -1 for non-OS errors and keep error codes unsigned?

Actually I wonder if it would be better to have two separate error domains: one for "well-known", OS-independent I/O errors such as `ErrorKind::NotFound`, and another one for everything else, which is going to be OS-specific. With this separation, user code would be able to do something like

```
open_file(path, &err);
if (err != NULL &&
    error_get_domain(err) == IO_ERROR &&
    error_get_code(err) == IO_ERROR_NOT_FOUND) {
    create_default_file(path);
}
```

This approach is similar to what GLib does with `GIOErrorEnum`. They don't even seem to have a way to obtain the raw `errno` value, so perhaps we could get away without it as well?

Reply:
I, for one, think that using error codes for choosing code paths is defective by design. If the error is something that cannot be checked for before the operation (requires atomicity, e.g. file exists/open) than the API should provide a way to do different things. The error should be used only for reporting error. Of course you might get to a place where you are forwarding an error from a place where this could happen and had to be done that way, however given the fact that VirtBlocks is a bit higher level it can be worked around in here instead of in all callers. Just my 2 cents.

Reply:
errno are not always >= 1. (naive grep "pub const E.* = -" in libc create). I think using signed int gives more latitude. Mapping all errno to a different (sub)set (like glib) is imho not very useful, but can be done.

Reply:
> errno are not always >= 1. (naive grep "pub const E.* = -" in libc create).

```
libc-0.2.62$ grep -rE 'pub const E.* = -.' | grep -vE 'EOF|EAI_|EVFILT_' | sed -E 's/:.*//g' | sort -u
src/unix/haiku/mod.rs
```

Once you remove `EOF`, which is obviously not an `errno`, the various `EAI`s which are return values for `getaddrinfo()` and the various `EVFILT`s which are used by BSD's `kqueue()`, you're left with a bunch of stuff from Haiku. I hope nobody takes this personally, but we don't care about Haiku O:-)

So for all intents and purposes `errno` values are positive, non zero integers, and we can represent them as such.

> I think using signed int gives more latitude. Mapping all errno to a different (sub)set (like glib) is imho not very useful, but can be done.

It seems that both GLib and Rust's own stdlib agree that users mostly care about a relatively small subset of error codes, but I'm also okay with exposing `errno`s directly and ask users to compare the return value of `virtblocks_error_get_code()` with the various constants defined in `errno.h`: it's actually less work for us right now, and we can always change it later if we change our minds.